File: app/src/main/java/com/introduce/hotel/view/ProfileFragment.java
package com.introduce.hotel.view;

import static android.app.Activity.RESULT_OK;

import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.UserProfileChangeRequest;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;

import java.io.IOException;

import com.introduce.hotel.R;

public class ProfileFragment extends Fragment {

    private static final int PICK_IMAGE_REQUEST = 71;
    private ImageView imgProfile;
    private Uri filePath;
    private FirebaseAuth mAuth;
    private FirebaseStorage storage;
    private StorageReference storageReference;
    private EditText edtName;
    private TextView txtName;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_profile, container, false);

        mAuth = FirebaseAuth.getInstance();
        storage = FirebaseStorage.getInstance();
        storageReference = storage.getReference();

        imgProfile = view.findViewById(R.id.profile_image);
        Button btnChooseImage = view.findViewById(R.id.edit_profile_image_button);
        Button btnUpdate = view.findViewById(R.id.edit_profile_name_button);
        edtName = view.findViewById(R.id.profile_name_edittext); // EditText for name input
        txtName = view.findViewById(R.id.profile_name); // TextView for displaying name

        // Initialize with current user info
        FirebaseUser user = mAuth.getCurrentUser();
        if (user != null) {
            if (user.getPhotoUrl() != null) {
                imgProfile.setImageURI(user.getPhotoUrl()); // Set the user's profile image if exists
            }
            txtName.setText(user.getDisplayName()); // Set the user's name
        }

        btnChooseImage.setOnClickListener(v -> chooseImage());

        btnUpdate.setOnClickListener(v -> updateProfile());

        return view;
    }

    private void chooseImage() {
        Intent intent = new Intent();
        intent.setType("image/*");
        intent.setAction(Intent.ACTION_GET_CONTENT);
        startActivityForResult(Intent.createChooser(intent, "Select Picture"), PICK_IMAGE_REQUEST);
    }

    private void updateProfile() {
        String name = edtName.getText().toString().trim();
        if (filePath != null) {
            StorageReference ref = storageReference.child("profileImages/" + mAuth.getCurrentUser().getUid());
            ref.putFile(filePath).addOnSuccessListener(taskSnapshot -> ref.getDownloadUrl().addOnSuccessListener(uri -> {
                UserProfileChangeRequest profileUpdate = new UserProfileChangeRequest.Builder()
                        .setDisplayName(name)
                        .setPhotoUri(uri)
                        .build();
                mAuth.getCurrentUser().updateProfile(profileUpdate)
                        .addOnCompleteListener(task -> {
                            if (task.isSuccessful()) {
                                Toast.makeText(getContext(), "Profile Updated", Toast.LENGTH_SHORT).show();
                                txtName.setText(name); // Update displayed name
                            }
                        });
            })).addOnFailureListener(e -> Toast.makeText(getContext(), "Failed to upload image", Toast.LENGTH_SHORT).show());
        } else {
            // Update name only
            UserProfileChangeRequest profileUpdate = new UserProfileChangeRequest.Builder()
                    .setDisplayName(name)
                    .build();
            mAuth.getCurrentUser().updateProfile(profileUpdate)
                    .addOnCompleteListener(task -> {
                        if (task.isSuccessful()) {
                            Toast.makeText(getContext(), "Profile Updated", Toast.LENGTH_SHORT).show();
                            txtName.setText(name); // Update displayed name
                        }
                    });
        }
    }

    @Override
    public  void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == PICK_IMAGE_REQUEST && resultCode == RESULT_OK && data != null && data.getData() != null) {
            filePath = data.getData();
            try {
                Bitmap bitmap = MediaStore.Images.Media.getBitmap(getContext().getContentResolver(), filePath);
                imgProfile.setImageBitmap(bitmap);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}

File: app/src/main/java/com/introduce/hotel/view/ProfileFragment.kt
package com.introduce.hotel.view

import android.Manifest
import android.app.Activity
import android.app.Activity.RESULT_OK
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResultCallback
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.UserProfileChangeRequest
import com.google.firebase.storage.FirebaseStorage
import com.google.firebase.storage.StorageReference
import com.introduce.hotel.R
import com.squareup.picasso.Picasso

class ProfileFragment : Fragment() {

    private lateinit var imgProfile: ImageView
    private var filePath: Uri? = null
    private lateinit var mAuth: FirebaseAuth
    private lateinit var storage: FirebaseStorage
    private lateinit var storageReference: StorageReference
    private lateinit var edtName: EditText
    private var selectedImageUri: Uri? = null
    companion object {
        private const val REQUEST_CODE_GALLERY = 1
    }
    private val imagePickerLauncher = registerForActivityResult<String, Uri>(
        ActivityResultContracts.GetContent(),
        object : ActivityResultCallback<Uri?> {
            override fun onActivityResult(result: Uri?) {
                if (result != null) {
                    try {
                        result?.let {
                            Picasso.get().load(it).into(imgProfile)
                            selectedImageUri = it
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
            }
        }
    )
    override fun onCreateView(
            inflater: LayoutInflater, container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View? {
        val view = inflater.inflate(R.layout.fragment_profile, container, false)

        mAuth = FirebaseAuth.getInstance()
        storage = FirebaseStorage.getInstance()
        storageReference = storage.reference

        imgProfile = view.findViewById(R.id.profile_image)
        val btnChooseImage = view.findViewById<Button>(R.id.edit_profile_image_button)
        val btnUpdate = view.findViewById<Button>(R.id.edit_profile_name_button)
        val btnLogout = view.findViewById<Button>(R.id.logout_button)
        edtName = view.findViewById(R.id.profile_name_edittext)

        // Initialize with current user info
        val user = mAuth.currentUser
        user?.let {
            if (user.photoUrl != null) {
                Picasso.get().load(user.photoUrl).into(imgProfile)
            }
            edtName.setText(user.displayName)
        }

        btnChooseImage.setOnClickListener { chooseImage() }

        btnUpdate.setOnClickListener { updateProfile() }

        btnLogout.setOnClickListener { logout() }

        return view
    }

    private fun chooseImage() {
        if (ContextCompat.checkSelfPermission(requireContext(), Manifest.permission.READ_EXTERNAL_STORAGE)
            != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(
                requireContext() as Activity, arrayOf<String>(Manifest.permission.READ_EXTERNAL_STORAGE),
                REQUEST_CODE_GALLERY
            )
        } else {
            imagePickerLauncher.launch("image/*")
        }
    }
    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<String?>,
        grantResults: IntArray
    ) {
        //here i am checking if the phone has granted permission to the app to access gallery
        if (requestCode == REQUEST_CODE_GALLERY) {
            imagePickerLauncher.launch("image/*")
            return
        }
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
    }

    private fun updateProfile() {
        val name = edtName.text.toString().trim()
        filePath?.let { fileUri ->
            val ref = storageReference.child("profileImages/${mAuth.currentUser?.uid}")
            ref.putFile(fileUri).addOnSuccessListener { taskSnapshot ->
                ref.downloadUrl.addOnSuccessListener { uri ->
                    val profileUpdate = UserProfileChangeRequest.Builder()
                            .setDisplayName(name)
                            .setPhotoUri(uri)
                            .build()
                    mAuth.currentUser?.updateProfile(profileUpdate)?.addOnCompleteListener { task ->
                        if (task.isSuccessful) {
                            Toast.makeText(context, "Profile Updated", Toast.LENGTH_SHORT).show()
                        }
                    }
                    Picasso.get().load(uri).into(imgProfile)
                }
            }.addOnFailureListener {
                Toast.makeText(context, "Failed to upload image", Toast.LENGTH_SHORT).show()
            }
        } ?: run {
            // Update name only
            val profileUpdate = UserProfileChangeRequest.Builder()
                    .setDisplayName(name)
                    .build()
            mAuth.currentUser?.updateProfile(profileUpdate)?.addOnCompleteListener { task ->
                if (task.isSuccessful) {
                    Toast.makeText(context, "Profile Updated", Toast.LENGTH_SHORT).show()
                }
            }
        }
    }

    private fun logout() {
        mAuth.signOut()
        startActivity(Intent(context, LoginActivity::class.java))
        activity?.finish()
    }
}
